src/serac/numerics/functional/functional.hpp

```diff
@@ -25,6 +25,54 @@
 
 namespace serac {
 
+// this type exists as a way to signal to `serac::Functional` that the function
+// serac::Functional::operator()` should differentiate w.r.t. a specific argument
+//
+// TODO: better name
+struct dual_vector {
+  const mfem::Vector& ref;
+
+  operator const mfem::Vector &() const { return ref; }
+};
+
+/**
+ * @brief this function is intended to only be used in combination with
+ *   `serac::Functional::operator()`, as a way for the user to express that
+ *   it should both evaluate and differentiate w.r.t. a specific argument (only 1 argument at a time)
+ *
+ * For example:
+ * @code{.cpp}
+ *     mfem::Vector arg0 = ...;
+ *     mfem::Vector arg1 = ...;
+ *     mfem::Vector just_the_value = my_functional(arg0, arg1);
+ *     auto [value, gradient_wrt_arg1] = my_functional(arg0, differentiate_wrt(arg1));
+ * @endcode
+ */
+dual_vector differentiate_wrt(const mfem::Vector& v) { return dual_vector{v}; }
+
+/**
+ * @tparam T a list of types, containing at most 1 `dual_vector`
+ *
+ * @brief given a list of types, this function returns the index that corresponds to the type `dual_vector`.
```

Comment:
```suggestion
 * @brief given a list of types, this function returns the index that corresponds to the type `dual_vector`.
 *
 * @tparam T a list of types, containing at most 1 `dual_vector`
```

Reply:
Does this change the doxygen output, or is it a stylistic choice? 

If it's a style thing, maybe we could put the expected doxygen comment format [here](https://serac.readthedocs.io/en/latest/sphinx/dev_guide/style_guide.html#documentation).

Reply:
Style/consistency.  We can handle it after this PR.

#629